read nearest neighbor 2d

SuperOptimizer opened the pull request:
single threaded since the current `readinterpolated3d` fails to compile on my machine:

```
s shape [4889,1775,1700]
ds shape via chunk [4889,1775,1700]
chunk shape shape [128,128,128]
testing different slice directions / caching w/ trilinear interpolation
0.00380702s gen_coords() yz cold
0.0174932s slicing 3d trilinear interpolation / 57.1651MiB/s yz cold
0.00327957s gen_coords() yz
0.0169649s slicing 3d trilinear interpolation / 58.9453MiB/s yz
0.00330981s gen_coords() xz cold
0.017207s slicing 3d trilinear interpolation / 58.1158MiB/s xz cold
0.00135848s gen_coords() xz
0.0169774s slicing 3d trilinear interpolation / 58.902MiB/s xz
0.00135877s gen_coords() xy cold
0.017039s slicing 3d trilinear interpolation / 58.689MiB/s xy cold
0.00135386s gen_coords() xy
0.0168931s slicing 3d trilinear interpolation / 59.1958MiB/s xy
0.00136025s gen_coords() diag cold
0.0174454s slicing 3d trilinear interpolation / 57.3219MiB/s diag cold
0.00135114s gen_coords() diag
0.017106s slicing 3d trilinear interpolation / 58.4591MiB/s diag
0.00135017s gen_coords() yz cold
0.0168361s slicing 3d trilinear interpolation / 59.3962MiB/s yz cold
0.00135554s gen_coords() yz
0.0169519s slicing 3d trilinear interpolation / 58.9905MiB/s yz
0.00136341s gen_coords() xz cold
0.0167511s slicing 3d trilinear interpolation / 59.6975MiB/s xz cold
0.00135322s gen_coords() xz
0.0166995s slicing 3d trilinear interpolation / 59.882MiB/s xz
0.0013526s gen_coords() xy cold
0.0168575s slicing 3d trilinear interpolation / 59.3209MiB/s xy cold
0.0013584s gen_coords() xy
0.016053s slicing 3d trilinear interpolation / 62.2938MiB/s xy
0.00119797s gen_coords() diag cold
0.0140453s slicing 3d trilinear interpolation / 71.1983MiB/s diag cold
0.00104145s gen_coords() diag
0.0123873s slicing 3d trilinear interpolation / 80.7279MiB/s diag
0.000954879s gen_coords() yz cold
0.0117806s slicing 3d trilinear interpolation / 84.8856MiB/s yz cold
0.000970869s gen_coords() yz
0.0122591s slicing 3d trilinear interpolation / 81.5721MiB/s yz
0.00096597s gen_coords() xz cold
0.0118239s slicing 3d trilinear interpolation / 84.5746MiB/s xz cold
0.000960169s gen_coords() xz
0.0118529s slicing 3d trilinear interpolation / 84.3673MiB/s xz
0.000966581s gen_coords() xy cold
0.0118527s slicing 3d trilinear interpolation / 84.3687MiB/s xy cold
0.000970449s gen_coords() xy
0.0118502s slicing 3d trilinear interpolation / 84.3869MiB/s xy
0.000968354s gen_coords() diag cold
0.0121603s slicing 3d trilinear interpolation / 82.2346MiB/s diag cold
0.000965519s gen_coords() diag
0.0121257s slicing 3d trilinear interpolation / 82.4697MiB/s diag
testing different slice directions / caching w/ nearest neighbor
0.000966982s gen_coords() yz cold
0.00203978s slicing 2d nearest neighbor / 490.248MiB/s yz cold
0.00096566s gen_coords() yz
0.00183039s slicing 2d nearest neighbor / 546.332MiB/s yz
0.0009657s gen_coords() xz cold
0.00237784s slicing 2d nearest neighbor / 420.55MiB/s xz cold
0.00100455s gen_coords() xz
0.00253953s slicing 2d nearest neighbor / 393.774MiB/s xz
0.00098199s gen_coords() xy cold
0.0019405s slicing 2d nearest neighbor / 515.332MiB/s xy cold
0.000967573s gen_coords() xy
0.00182307s slicing 2d nearest neighbor / 548.526MiB/s xy
0.000969847s gen_coords() diag cold
0.00231163s slicing 2d nearest neighbor / 432.595MiB/s diag cold
0.000969737s gen_coords() diag
0.00200569s slicing 2d nearest neighbor / 498.582MiB/s diag
0.000987481s gen_coords() yz cold
0.00229825s slicing 2d nearest neighbor / 435.114MiB/s yz cold
0.000971311s gen_coords() yz
0.00186305s slicing 2d nearest neighbor / 536.754MiB/s yz
0.00097118s gen_coords() xz cold
0.00180903s slicing 2d nearest neighbor / 552.782MiB/s xz cold
0.000969607s gen_coords() xz
0.00177571s slicing 2d nearest neighbor / 563.156MiB/s xz
0.000971381s gen_coords() xy cold
0.00177461s slicing 2d nearest neighbor / 563.505MiB/s xy cold
0.000969757s gen_coords() xy
0.00205687s slicing 2d nearest neighbor / 486.177MiB/s xy
0.000979035s gen_coords() diag cold
0.00220452s slicing 2d nearest neighbor / 453.613MiB/s diag cold
0.000975538s gen_coords() diag
0.00197928s slicing 2d nearest neighbor / 505.234MiB/s diag
0.000981459s gen_coords() yz cold
0.00183822s slicing 2d nearest neighbor / 544.003MiB/s yz cold
0.00102523s gen_coords() yz
0.00207303s slicing 2d nearest neighbor / 482.385MiB/s yz
0.000973624s gen_coords() xz cold
0.00184096s slicing 2d nearest neighbor / 543.195MiB/s xz cold
0.000983103s gen_coords() xz
0.0023316s slicing 2d nearest neighbor / 428.89MiB/s xz
0.000975589s gen_coords() xy cold
0.00189687s slicing 2d nearest neighbor / 527.183MiB/s xy cold
0.000972422s gen_coords() xy
0.00181515s slicing 2d nearest neighbor / 550.918MiB/s xy
0.000972893s gen_coords() diag cold
0.00193398s slicing 2d nearest neighbor / 517.067MiB/s diag cold
0.000973254s gen_coords() diag
0.0019101s slicing 2d nearest neighbor / 523.533MiB/s diag
1.29197s slicing / 77.4012MiB/s  shift (cold)
1.29376s slicing / 77.2938MiB/s  shift (warm)
(base) forrest@forlinlap ~/CLionProjects/volume-cartographer/cmake-build-release/bin (readnearestneighbor) $ 



```

SuperOptimizer committed:
read nearest neighbor 2d

SuperOptimizer committed:
read nearest neighbor 2d

SuperOptimizer committed:
remove #pragma in readinterpolated3d that was causing compile errors

SuperOptimizer commented:
I had to remove the `omp parallel` above the chunk getting step. it's spelled wrong now but when I fixed it I got other errors.

multithreaded:
```
(base) forrest@forlinlap ~/CLionProjects/volume-cartographer/cmake-build-release/bin (readnearestneighbor) $ OMP_NUM_THREADS=16 ./vc_render_bench ~/scroll3.volpkg/volumes/scroll3.zarr
ds shape [4889,1775,1700]
ds shape via chunk [4889,1775,1700]
chunk shape shape [128,128,128]
testing different slice directions / caching w/ trilinear interpolation
0.00322564s gen_coords() yz cold
0.0123277s slicing 3d trilinear interpolation / 81.1182MiB/s yz cold
0.00249329s gen_coords() yz
0.0118777s slicing 3d trilinear interpolation / 84.1911MiB/s yz
0.00240799s gen_coords() xz cold
0.012141s slicing 3d trilinear interpolation / 82.3654MiB/s xz cold
0.000959961s gen_coords() xz
0.0118091s slicing 3d trilinear interpolation / 84.6805MiB/s xz
0.00095934s gen_coords() xy cold
0.0120237s slicing 3d trilinear interpolation / 83.1689MiB/s xy cold
0.000962746s gen_coords() xy
0.0119349s slicing 3d trilinear interpolation / 83.7882MiB/s xy
0.000965962s gen_coords() diag cold
0.0123732s slicing 3d trilinear interpolation / 80.8196MiB/s diag cold
0.00096491s gen_coords() diag
0.0121448s slicing 3d trilinear interpolation / 82.3399MiB/s diag
0.000963016s gen_coords() yz cold
0.0118505s slicing 3d trilinear interpolation / 84.3844MiB/s yz cold
0.000985339s gen_coords() yz
0.0118411s slicing 3d trilinear interpolation / 84.4517MiB/s yz
0.000961564s gen_coords() xz cold
0.0118301s slicing 3d trilinear interpolation / 84.5305MiB/s xz cold
0.000961434s gen_coords() xz
0.0118469s slicing 3d trilinear interpolation / 84.4105MiB/s xz
0.000964389s gen_coords() xy cold
0.0118538s slicing 3d trilinear interpolation / 84.3609MiB/s xy cold
0.000983786s gen_coords() xy
0.0118401s slicing 3d trilinear interpolation / 84.4585MiB/s xy
0.000966694s gen_coords() diag cold
0.0121493s slicing 3d trilinear interpolation / 82.3091MiB/s diag cold
0.000963939s gen_coords() diag
0.0121336s slicing 3d trilinear interpolation / 82.4154MiB/s diag
0.000965511s gen_coords() yz cold
0.0118879s slicing 3d trilinear interpolation / 84.119MiB/s yz cold
0.000965993s gen_coords() yz
0.0118265s slicing 3d trilinear interpolation / 84.5555MiB/s yz
0.000998844s gen_coords() xz cold
0.0118732s slicing 3d trilinear interpolation / 84.2234MiB/s xz cold
0.000963588s gen_coords() xz
0.0118488s slicing 3d trilinear interpolation / 84.3968MiB/s xz
0.000965401s gen_coords() xy cold
0.0118497s slicing 3d trilinear interpolation / 84.3904MiB/s xy cold
0.000965621s gen_coords() xy
0.011856s slicing 3d trilinear interpolation / 84.3456MiB/s xy
0.00098045s gen_coords() diag cold
0.0121451s slicing 3d trilinear interpolation / 82.3374MiB/s diag cold
0.000964419s gen_coords() diag
0.0121631s slicing 3d trilinear interpolation / 82.2161MiB/s diag
testing different slice directions / caching w/ nearest neighbor
0.000966864s gen_coords() yz cold
0.00209451s slicing 2d nearest neighbor / 477.438MiB/s yz cold
0.000963607s gen_coords() yz
0.00204576s slicing 2d nearest neighbor / 488.816MiB/s yz
0.000977634s gen_coords() xz cold
0.00273536s slicing 2d nearest neighbor / 365.583MiB/s xz cold
0.000982533s gen_coords() xz
0.00189689s slicing 2d nearest neighbor / 527.179MiB/s xz
0.000967555s gen_coords() xy cold
0.00187075s slicing 2d nearest neighbor / 534.545MiB/s xy cold
0.000968436s gen_coords() xy
0.00176227s slicing 2d nearest neighbor / 567.448MiB/s xy
0.000969299s gen_coords() diag cold
0.00222302s slicing 2d nearest neighbor / 449.838MiB/s diag cold
0.000982102s gen_coords() diag
0.00246751s slicing 2d nearest neighbor / 405.266MiB/s diag
0.000975039s gen_coords() yz cold
0.00190285s slicing 2d nearest neighbor / 525.528MiB/s yz cold
0.00097005s gen_coords() yz
0.00182672s slicing 2d nearest neighbor / 547.43MiB/s yz
0.000984527s gen_coords() xz cold
0.00178867s slicing 2d nearest neighbor / 559.073MiB/s xz cold
0.000969509s gen_coords() xz
0.00178094s slicing 2d nearest neighbor / 561.502MiB/s xz
0.000969919s gen_coords() xy cold
0.00203373s slicing 2d nearest neighbor / 491.708MiB/s xy cold
0.000995167s gen_coords() xy
0.00214225s slicing 2d nearest neighbor / 466.799MiB/s xy
0.000974778s gen_coords() diag cold
0.00205158s slicing 2d nearest neighbor / 487.429MiB/s diag cold
0.000974148s gen_coords() diag
0.00196643s slicing 2d nearest neighbor / 508.536MiB/s diag
0.000976662s gen_coords() yz cold
0.00182148s slicing 2d nearest neighbor / 549.005MiB/s yz cold
0.000971713s gen_coords() yz
0.00176948s slicing 2d nearest neighbor / 565.138MiB/s yz
0.00100649s gen_coords() xz cold
0.00230378s slicing 2d nearest neighbor / 434.07MiB/s xz cold
0.000974458s gen_coords() xz
0.00190047s slicing 2d nearest neighbor / 526.187MiB/s xz
0.00097009s gen_coords() xy cold
0.00177793s slicing 2d nearest neighbor / 562.451MiB/s xy cold
0.000970691s gen_coords() xy
0.00179956s slicing 2d nearest neighbor / 555.69MiB/s xy
0.000972554s gen_coords() diag cold
0.00190686s slicing 2d nearest neighbor / 524.423MiB/s diag cold
0.000971663s gen_coords() diag
0.00191335s slicing 2d nearest neighbor / 522.644MiB/s diag
1.29535s slicing / 77.1991MiB/s  shift (cold)
1.29866s slicing / 77.0023MiB/s  shift (warm)
(base) forrest@forlinlap ~/CLionProjects/volume-cartographer/cmake-build-release/bin (readnearestneighbor) $ 
```

hendrikschilling commented:
Are you compiling without openmp? I'm getting some wildly different performance numbers.
Also please make this configurable in VC3D:
- add a flag to readinterpolated3d (linear/nearest) with default linear to stay backwards compatible (and don't expose the NN implementation by itself)
- add a config option in VC3D gui to change the flag to nearest
And please don't call it 2D, its still interpolating in 3d, how else would it read from a volume :-P

hendrikschilling commented:
Ah one more thing - the bench tool should get flag too and otherwise run the same code - the latency testing (cold) can not work if its run one after the other.

hendrikschilling commented:
as the code doesn't support non-power-of-two chunk sizes in that case it should probably throw an exception.

SuperOptimizer committed:
basic threading for nearest neighbor

SuperOptimizer committed:
make nearest neighbor configurable by settings

SuperOptimizer committed:
Merge branch 'dev-next' into readnearestneighbor

SuperOptimizer committed:
fix merge bug

SuperOptimizer commented:
> Are you compiling without openmp? I'm getting some wildly different performance numbers. Also please make this configurable in VC3D:
> 
>     * add a flag to readinterpolated3d (linear/nearest) with default linear to stay backwards compatible (and don't expose the NN implementation by itself)
> 
>     * add a config option in VC3D gui to change the flag to nearest
>       And please don't call it 2D, its still interpolating in 3d, how else would it read from a volume :-P

turns out the openmp perf issue was related to using llvm + clang + llvm's libomp. switching to gcc fixed that; though AFAICT llvm+clang outputs a faster single threaded version :upside_down_face: I've added a settings switch and renamed it to just `readnearestneighbor`

> Ah one more thing - the bench tool should get flag too and otherwise run the same code - the latency testing (cold) can not work if its run one after the other.

done



> as the code doesn't support non-power-of-two chunk sizes in that case it should probably throw an exception.

done
